kernel/tclapi.cc

```diff
@@ -0,0 +1,449 @@
+	} else if (false_flag) {
+		obj->set_bool_attribute(attr_id, false);
+	} else {
+		obj->attributes[attr_id] = Const::from_string(std::string(argv[i++]));
+	}
+
+	return TCL_OK;
+}
+
+static int tcl_get_param(ClientData, Tcl_Interp *interp, int argc, const char *argv[])
+{
+	int i;
+	bool string_flag = false, int_flag = false;
+	for (i = 1; i < argc; i++) {
+		FLAG(string)
+		FLAG(int)
+		break;
+	}
+
+	if ((i != argc - 3) ||
+			(string_flag + int_flag > 1))
+		ERROR("bad usage: expected \"get_param [-string|-int] <module> <cellid> <paramname>")
+
+	IdString mod_id, cell_id, param_id;
+	mod_id = RTLIL::escape_id(argv[i++]);
+	cell_id = RTLIL::escape_id(argv[i++]);
+	param_id = RTLIL::escape_id(argv[i++]);
+
+	RTLIL::Module *mod = yosys_design->module(mod_id);
+	if (!mod)
+		ERROR("module not found")
+
+	RTLIL::Cell *cell = mod->cell(cell_id);
+	if (!cell)
+		ERROR("object not found")
+
+	if (!cell->hasParam(param_id))
+		ERROR("parameter missing")
+
+	const RTLIL::Const &value = cell->getParam(param_id);
+
+	if (string_flag) {
+		Tcl_SetResult(interp, (char *) value.decode_string().c_str(), TCL_VOLATILE);
+	} else if (int_flag) {
+		if (value.size() > 32)
+			ERROR("value too large")
+
+		Tcl_SetResult(interp, (char *) std::to_string(value.as_int()).c_str(), TCL_VOLATILE);
+	} else {
+		Tcl_SetResult(interp, (char *) value.as_string().c_str(), TCL_VOLATILE);
+	}
+	return TCL_OK;
+}
+
+static int tcl_set_param(ClientData, Tcl_Interp *interp, int argc, const char *argv[])
+{
+	int i;
+	bool string_flag = false, int_flag = false;
+	for (i = 1; i < argc; i++) {
+		FLAG(string)
+		FLAG(int)
+		break;
+	}
+
+	if ((i != argc - 4) ||
+			(string_flag + int_flag > 1))
+		ERROR("bad usage: expected \"get_param [-string|-int] <module> <cellid> <paramname> <value>")
+
+	IdString mod_id, cell_id, param_id;
+	mod_id = RTLIL::escape_id(argv[i++]);
+	cell_id = RTLIL::escape_id(argv[i++]);
+	param_id = RTLIL::escape_id(argv[i++]);
+
+	RTLIL::Module *mod = yosys_design->module(mod_id);
+	if (!mod)
+		ERROR("module not found")
+
+	RTLIL::Cell *cell = mod->cell(cell_id);
+	if (!cell)
+		ERROR("object not found")
+
+	if (string_flag) {
+		cell->setParam(param_id, Const(std::string(argv[i++])));
+	} else if (int_flag) {
+		cell->setParam(param_id, Const(atoi(argv[i++])));
+	} else {
+		cell->setParam(param_id, Const::from_string(std::string(argv[i++])));
+	}
+	return TCL_OK;
+}
+
+int yosys_tcl_iterp_init(Tcl_Interp *interp)
+{
+	if (Tcl_Init(interp)!=TCL_OK)
+		log_warning("Tcl_Init() call failed - %s\n",Tcl_ErrnoMsg(Tcl_GetErrno()));
+	Tcl_CreateCommand(interp, "yosys", tcl_yosys_cmd, NULL, NULL);
+	Tcl_CreateCommand(interp, "rtlil::get_attr", tcl_get_attr, NULL, NULL);
+	Tcl_CreateCommand(interp, "rtlil::has_attr", tcl_has_attr, NULL, NULL);
+	Tcl_CreateCommand(interp, "rtlil::set_attr", tcl_set_attr, NULL, NULL);
+	Tcl_CreateCommand(interp, "rtlil::get_param", tcl_get_param, NULL, NULL);
+	Tcl_CreateCommand(interp, "rtlil::set_param", tcl_set_param, NULL, NULL);
```

Comment:
I'd also expect `rtlil::unset_attr` as well as `rtlil::attr_list` to get a complete set of operations (same for param which is also missing `rtlil::has_param`).

Maybe also useful would be something like `rtlil::attr_array` which would produce all attributes as a key-value array such that you can do `array set object_attrs [rtlil::attr_array some_mod some_object]; puts "the value of keep is $object_attrs(\\keep)"` and use other tcl array functions. I don't actually use tcl enough to know how useful this would be, though.